File: scripts/vscripts/ViSDK/ViGameCommon/ViTickNode.ts
import { ViDoubleLink, ViDoubleLinkNode } from "../ViStruct/ViDoubleLink";
import { ViDelegate1, ViDelegater1 } from "../ViSystem/ViDelegate";
import { ViTimeNodeInterface } from "./Time/ViTimer";

export class ViTickNode
{
	private static readonly CACHE_Update_ExecList = new ViDoubleLink<ViTickNode>();
	public static Update(deltaTime: number): void
	{
		let _list = ViTickNode._list;
		if (_list.Empty)
		{
			return;
		}
		//
		let CACHE_Update_ExecList = ViTickNode.CACHE_Update_ExecList;
		while (_list.NotEmpty)
		{
			let iterNode = _list.GetHead();
			CACHE_Update_ExecList.PushBack(iterNode);
			let iterCallback = iterNode.Data;
			iterCallback.Invoke(deltaTime);
		}
		//
		_list.PushListBack(CACHE_Update_ExecList);
	}
	//
	public static Clear(): void
	{
		ViTickNode.CACHE_Update_ExecList.Clear();
		ViTickNode._list.Clear();
	}
	//
	private static readonly _list = new ViDoubleLink<ViTickNode>();
	//
	//+-------------------------------------------------------------------------------------------------------------------------------------------------------------
	public Attach(listener: any, func: ViDelegate1<number>): void
	{
		this._callback.Set(listener, func);
		ViTickNode._list.PushBackEx(this._node, this);
	}
	//
	public Detach(): void
	{
		this._callback.Clear();
		this._node.DetachEx(null);
	}
	//
	public IsAttach(): boolean
	{
		return this._node.IsAttach();
	}
	//
	public Invoke(deltaTime: number)
	{
		this._callback.Invoke(deltaTime);
	}
	//
	private readonly _node = new ViDoubleLinkNode<ViTickNode>();
	private readonly _callback = new ViDelegater1();
}

File: scripts/vscripts/Game/ViSDK/ViTime/ViTickNode.lua
local ViTickNode = BaseClass("ViTickNode")

local function ViTickNodeCtor(self)
    self._delegate = ViDelegate1.New()
    self._node = ViDoubleLinkNode2.New()
end

local function Update(deltaTime)
    local tickList = ViTickNode._tickList
    if tickList:IsEmpty() then
        return
    end
    local CACHE_UPDATE_ExecList = ViTickNode.CACHE_UPDATE_ExecList
    while tickList:IsNotEmpty() do
        local node = tickList:GetHead()
        CACHE_UPDATE_ExecList:PushBackNode(node)
        local tickNode = node.Data
        if tickNode ~= nil then
            tickNode:Exec(deltaTime)
        end
    end
    tickList:PushBackList(CACHE_UPDATE_ExecList)
end

local function Attach(self, listener, callback)
    self._delegate:SetDele(listener, callback)
    self._node.Data = self
    ViTickNode._tickList:PushBackNode(self._node)
end

local function IsAttach(self)
   return self._node:IsAttach()
end

local function Detach(self)
    ViDelegateAssisstant.Clear(self._delegate)
    self._node:DetachEX(nil)
end

local function Clear()
    ViTickNode._tickList:Clear()
    ViTickNode.CACHE_UPDATE_ExecList:Clear()
end

local function Exec(self, deltaTime)
    self._delegate:Invoke(deltaTime)
end

ViTickNode.__init = ViTickNodeCtor
ViTickNode.Update = Update
ViTickNode.Attach = Attach
ViTickNode.Detach = Detach
ViTickNode.IsAttach = IsAttach
ViTickNode.Exec = Exec

ViTickNode._tickList = ViDoubleLink2.New()
ViTickNode.CACHE_UPDATE_ExecList = ViDoubleLink2.New()

return ViTickNode